shm/src/ship.cpp

```diff
@@ -0,0 +1,89 @@
+#include "ship.hpp"
+
+#include <algorithm>
+#include <iostream>
+
+Ship::Ship() : id_(-1) {}
+
+Ship::Ship(int capacity,
+           int maxCrew,
+           int speed,
+           const std::string& name,
+           size_t id,
+           Delegate* delegate,
+           Time* time)
+    : delegate_(delegate),
+      capacity_(capacity),
+      maxCrew_(maxCrew),
+      crew_(0),
+      speed_(speed),
+      name_(name),
+      id_(id),
+      time_(time) {
+    time_->addObserver(this);
+}
+
+Ship::Ship(int maxCrew, int speed, size_t id, Delegate* delegate, Time* time)
+    : Ship(0, maxCrew, speed, "", id, delegate, time) {}
+
+void Ship::setName(const std::string& name) {
+    name_ = name;
+}
+
+Ship& Ship::operator-=(size_t crew) {
+    crew_ -= crew;
+    return *this;
+}
+Ship& Ship::operator+=(size_t crew) {
+    crew_ += crew;
+    return *this;
```

Comment:
Where are you checking if you have space for new crewmen?

Reply:
Hi - this is CMAKE related MR. We have another PR which will be resolved today,  so plase give comments to that.